File: prefix_sum/equal_array_min_cost/equal_array_min_cost.cpp
#include <vector>
#include <map>
using namespace std;

long long minCost(vector<int> &nums, vector<int> &costs)
{ // LeetCode Q.2448.
    map<int, long long> numsTotalCosts;

    for (int idx = 0; idx < nums.size(); idx++)
        numsTotalCosts[nums[idx]] += costs[idx];

    vector<int> distinctNums;
    // Weighted costs: cumulation of each num * its total costs.
    vector<long long> prefixCosts, prefixWeightedCosts;

    for (auto const &pair : numsTotalCosts)
    {
        distinctNums.push_back(pair.first);

        if (prefixCosts.empty())
            prefixCosts.push_back(0);

        else
            prefixCosts.push_back(prefixCosts.back());

        prefixCosts.back() += pair.second;

        if (prefixWeightedCosts.empty())
            prefixWeightedCosts.push_back(0);

        else
            prefixWeightedCosts.push_back(prefixWeightedCosts.back());

        prefixWeightedCosts.back() += pair.first * pair.second;
    }

    long long minCost = numeric_limits<long long>::max();

    for (int idx = 0; idx < distinctNums.size(); idx++)
    {
        long long num = distinctNums[idx];

        long long cost = prefixWeightedCosts.back() - prefixWeightedCosts[idx];
        cost -= num * (prefixCosts.back() - prefixCosts[idx]);

        if (idx > 0)
        {
            cost += num * (prefixCosts[idx - 1]);
            cost -= prefixWeightedCosts[idx - 1];
        }

        if (cost < minCost)
            minCost = cost;
    }

    return minCost;
}

File: prefix_sum/equal_array_min_cost/equal_array_min_cost.py

def calculate_min_cost(nums: list[int], costs: list[int]) -> int:  # LeetCode Q.2448.
    nums2total_costs: dict[int, int] = dict()
    for idx, num in enumerate(nums):
        if num not in nums2total_costs.keys():
            nums2total_costs.update({num: 0})
        nums2total_costs[num] += costs[idx]

    nums2total_costs = dict(sorted(nums2total_costs.items()))
    distinct_nums = list(nums2total_costs.keys())

    prefix_costs, prefix_total_costs = [], []
    for num, total_costs in nums2total_costs.items():
        if not prefix_costs:
            prefix_costs.append(0)

        else:
            prefix_costs.append(prefix_costs[-1])

        prefix_costs[-1] += total_costs

        if not prefix_total_costs:
            prefix_total_costs.append(0)

        else:
            prefix_total_costs.append(prefix_total_costs[-1])

        prefix_total_costs[-1] += num * total_costs

    min_cost = float("inf")
    for idx, num in enumerate(distinct_nums):
        total_cost = prefix_total_costs[-1] - prefix_total_costs[idx]
        total_cost -= num * (prefix_costs[-1] - prefix_costs[idx])

        if idx > 0:
            total_cost += num * prefix_costs[idx - 1]
            total_cost -= prefix_total_costs[idx - 1]

        if total_cost < min_cost:
            min_cost = total_cost

    return min_cost
